File: Top_100_liked_questions/Q2_E/Solution.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
var mergeTwoLists = function(list1, list2) {
    if(list1.next === undefined ) return list2;
    if(list2.next === undefined) return list1;
    let start =  list1;
    if(list1.val < list2.val)
    {
        list1 = list.next;
    }
    else
    {
        start = list2;
        list2=list2.next;
    }
    let temp = start;
    while(list1 && list2)
    {
        if(list1.val<list2.val){
            temp.next = list1;
            list1=list1.next;
        }
        else{
            temp.next = list2;
            list2=list2.next;
        }
        temp = temp.next;
    }
    if(list1) temp.next = list1;
    else if(list2) temp.next = list2;
    return start;
};

var mergerTwoList = function(list1,list2)
{
    if (!list1 || !list2) return list1 ? list1 : list2;
    
    let mergedList = new ListNode();
    let current = mergedList;
  
    while (list1 && list2) {
      if (list1.val < list2.val) {
        current.next = list1;
        list1 = list1.next;
      } else {
        current.next = list2;
        list2 = list2.next;
      }
      current = current.next;
    }
  
    current.next = list1 || list2;
    return mergedList.next;
}
var traverse = function(x)
{
    while(x)
    {
       console.log(x.val);
       x=x.next;
    }
}


l1 = {val:4,next:null}
l2 ={val:2,next:l1}
l3 ={val:1,next:l2}
traverse(l3)

console.log("--------------------------")
l1 = {val:4,next:null}
l2 ={val:3,next:l1}
l4  ={val:1,next:l2}
traverse(l4)

console.log("--------------------------")

 x = mergerTwoList(l3,l4);
 traverse(x)

File: Top_100_liked_questions/Q2_E/Solution.java
package Top_100_liked_questions.Q2_E;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */


class Solution {

    public static void main(String[] args) {
        ListNode l1 = new ListNode();
        ListNode l2 = new ListNode();
        ListNode s = null;

        l2.val = getRand();
        l1.val = getRand();
        ListNode x = l1;
        for (int i =0;i < 10;i++)
        {
            ListNode temp = new ListNode();
            temp.val = x.val + getRand();
            x.next = temp;
            x =temp;
        }

        ListNode.traverse(l1);

        x = l2;
        for (int i =0;i < 10;i++)
        {
            ListNode temp = new ListNode();
            temp.val =x.val + getRand();
            x.next = temp;
            x =temp;
        }
        ListNode.traverse(l2);


        s = mergeTwoLists(l1, l2);
        ListNode.traverse(s);


    }
    public static int getRand()
    {
        return (int) Math.round(Math.random()*10);
    }

    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        //Since max length of linked list is 50 recursion is a viable solution otherwise itrative method
        ListNode s = null;
        if(null == list1)
        {
            s = list2;
        }
        else if(null == list2)
        {
            s = list1;
        }
        else if(list1.val <= list2.val)
        {
            s = list1;
            list1.next = mergeTwoLists(list1.next,list2);
        }
        else if(list1.val > list2.val)
        {
            s = list2;
            list2.next =  mergeTwoLists(list1,list2.next);
        }
        return s;
    }
}